File: 2020/day7/day7.py
#!/usr/bin/env python3
# https://adventofcode.com/2020/day/7

from pathlib import Path
from copy import deepcopy
from collections import defaultdict, deque, namedtuple
import sys, re, math, itertools, time
from functools import reduce
import re

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
lines = data.splitlines()


bags = {}

for line in lines:
    bag, contain = re.match(r"^(.+) bags contain (.+)\.$", line).groups()

    contained = {}
    for s in contain.split(","):
        if s == "no other bags":
            continue
        n, sub_bag = re.match(r"^\s?(\d+) (.+) bags?$", s).groups()
        contained[sub_bag] = int(n)

    bags[bag] = contained


# part 1

def contains_color(bag, color):
    for sub_bag in bags[bag]:
        if sub_bag == color:
            return True
        if contains_color(sub_bag, color) == True:
            return True
    return False


print(sum(contains_color(bag, "shiny gold") for bag in bags))


# part 2

def count(bag):
    t = 0
    for s, n in bags[bag].items():
        t += (count(s) + 1) * n
    return t


print(count("shiny gold"))

File: 2020/day7/day7.rs
//! [Day 7: Handy Haversacks](https://adventofcode.com/2020/day/7)

use regex::Regex;
use std::collections::HashMap;

struct Puzzle {
    // data: String,
    bags: HashMap<String, HashMap<String, u32>>,
}

impl Puzzle {
    fn new() -> Self {
        Self {
            bags: HashMap::new(),
        }
    }

    /// Get the puzzle input.
    fn configure(&mut self, data: &str) {
        // self.data = data;

        let pat = Regex::new(r"^\s?(\d+) (.+) bags?$").unwrap();

        for line in data.lines() {
            let (bag, contain) = line.split_once(" bags contain ").unwrap();

            let contain = contain.strip_suffix('.').unwrap();

            let mut contained = HashMap::new();

            for s in contain.split(',') {
                if s == "no other bags" {
                    continue;
                }

                let caps = pat.captures(s).unwrap();

                let n: u32 = caps[1].parse().unwrap();
                let sub_bag = &caps[2];

                contained.insert(sub_bag.to_string(), n);
            }

            self.bags.insert(bag.to_string(), contained);
        }
    }

    fn contains_color(&self, bag: &str, color: &str) -> bool {
        for sub_bag in self.bags[bag].keys() {
            if sub_bag == color || self.contains_color(sub_bag, color) {
                return true;
            }
        }

        false
    }

    fn count_bag(&self, bag: &str) -> u32 {
        self.bags[bag]
            .iter()
            .fold(0, |acc, (b, n)| acc + (self.count_bag(b) + 1) * n)
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        self.bags
            .keys()
            .map(|bag| u32::from(self.contains_color(bag, "shiny gold")))
            .sum()
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        self.count_bag("shiny gold")
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_part1() {
        let mut puzzle = Puzzle::new();
        puzzle.configure(&aoc::load_input_data("sample_1.txt"));
        assert_eq!(puzzle.part1(), 4);
    }

    #[test]
    fn test_part2() {
        let mut puzzle = Puzzle::new();
        puzzle.configure(&aoc::load_input_data("sample_2.txt"));
        assert_eq!(puzzle.part2(), 126);
    }
}
